### ISSUE TITLE
RNN training is very slow on Intel xpu

### ISSUE THREAD

**Issue by JimmysAIPG** [pytorch/pytorch#150254]
### 🐛 Describe the bug

Hello, I am a newbie and I just switched from the cuda environment to the xpu environment for learning. I found that when I use GRU or LSTM to train the model, the speed is very slow in the xpu environment. Is there a problem?

Please refer to the attachment for the code

[imdb.txt](https://github.com/user-attachments/files/19516292/imdb.txt)

### Versions

I am using the docker environment：
REPOSITORY                          TAG             IMAGE ID       CREATED        SIZE
intel/intel-extension-for-pytorch   2.6.10-xpu      ed87f8f3c7e0   3 weeks ago    15.1GB

root@97f8d2344ae3:/home# python collect_env.py 
[W329 08:10:54.485931320 OperatorEntry.cpp:154] Warning: Warning only once for all operators,  other operators may also be overridden.
  Overriding a previously registered kernel for the same operator and the same dispatch key
  operator: aten::_validate_compressed_sparse_indices(bool is_crow, Tensor compressed_idx, Tensor plain_idx, int cdim, int dim, int nnz) -> ()
    registered at /pytorch/build/aten/src/ATen/RegisterSchema.cpp:6
  dispatch key: XPU
  previous kernel: registered at /pytorch/build/aten/src/ATen/RegisterCPU.cpp:30477
       new kernel: registered at /build/intel-pytorch-extension/build/Release/csrc/gpu/csrc/aten/generated/ATen/RegisterXPU.cpp:468 (function operator())
Collecting environment information...
PyTorch version: 2.6.0+xpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.12.20-061220-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               8
On-line CPU(s) list:                  0-7
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Core(TM) i3-10100F CPU @ 3.60GHz
CPU family:                           6
Model:                                165
Thread(s) per core:                   2
Core(s) per socket:                   4
Socket(s):                            1
Stepping:                             3
CPU max MHz:                          4300.0000
CPU min MHz:                          800.0000
BogoMIPS:                             7200.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts vnmi md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            128 KiB (4 instances)
L1i cache:                            128 KiB (4 instances)
L2 cache:                             1 MiB (4 instances)
L3 cache:                             6 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-7
Vulnerability Gather data sampling:   Vulnerable
Vulnerability Itlb multihit:          KVM: Mitigation: Split huge pages
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Mitigation; Microcode
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] intel_extension_for_pytorch==2.6.10+xpu
[pip3] numpy==2.2.3
[pip3] pytorch-triton-xpu==3.2.0
[pip3] torch==2.6.0+xpu
[pip3] torchaudio==2.6.0+xpu
[pip3] torchvision==0.21.0+xpu
[conda] Could not collect
root@97f8d2344ae3:/home# [W329 08:10:57.987819736 OperatorEntry.cpp:154] Warning: Warning only once for all operators,  other operators may also be overridden.
  Overriding a previously registered kernel for the same operator and the same dispatch key
  operator: aten::_validate_compressed_sparse_indices(bool is_crow, Tensor compressed_idx, Tensor plain_idx, int cdim, int dim, int nnz) -> ()
    registered at /pytorch/build/aten/src/ATen/RegisterSchema.cpp:6
  dispatch key: XPU
  previous kernel: registered at /pytorch/build/aten/src/ATen/RegisterCPU.cpp:30477
       new kernel: registered at /build/intel-pytorch-extension/build/Release/csrc/gpu/csrc/aten/generated/ATen/RegisterXPU.cpp:468 (function operator())

cc @gujinghui @EikanWang @fengyuan14 @guangyey

**Comment by guangyey** [comment]
@JimmysAIPG May I know what GPU you are using?
```python
import torch
for i in range(torch.xpu.device_count()):
    print(torch.xpu.get_device_properties(i))
```

**Comment by MadamePingShockwave** [comment]
> RNN related OPs are not optimized on XPU backend, so it's expected to be slow

I found an issue raised by you in the xpu-ops repo, which aims to integrate the oneDNN implementation of RNN and is in the pt2.8 milestone. Does that mean we may expect acceleration for RNNs in pytorch 2.8? Do we have other common NN achitectures that are not optimized in the xpu backend?

**Comment by jianyizh** [comment]
> > RNN related OPs are not optimized on XPU backend, so it's expected to be slow
> 
> I found an issue raised by you in the xpu-ops repo, which aims to integrate the oneDNN implementation of RNN and is in the pt2.8 milestone. Does that mean we may expect acceleration for RNNs in pytorch 2.8? Do we have other common NN achitectures that are not optimized in the xpu backend?

No, We currently have no plans on RNN. If you want peak performance, try https://github.com/intel/intel-extension-for-pytorch/tree/xpu-main

**Comment by JimmysAIPG** [comment]
> [@JimmysAIPG](https://github.com/JimmysAIPG) May I know what GPU you are using?
> 
> import torch
> for i in range(torch.xpu.device_count()):
>     print(torch.xpu.get_device_properties(i))

@guangyey Sorry for the late reply, the graphics card I am using now is Intel Arc A770.
`_XpuDeviceProperties(name='Intel(R) Arc(TM) A770 Graphics', platform_name='Intel(R) Level-Zero', type='gpu', driver_version='1.3.32413', total_memory=15930MB, max_compute_units=512, gpu_eu_count=512, gpu_subslice_count=64, max_work_group_size=1024, max_num_sub_groups=128, sub_group_sizes=[8 16 32], has_fp16=1, has_fp64=0, has_atomic64=1)`

**Comment by JimmysAIPG** [comment]
> > > RNN related OPs are not optimized on XPU backend, so it's expected to be slow
> > 
> > 
> > I found an issue raised by you in the xpu-ops repo, which aims to integrate the oneDNN implementation of RNN and is in the pt2.8 milestone. Does that mean we may expect acceleration for RNNs in pytorch 2.8? Do we have other common NN achitectures that are not optimized in the xpu backend?
> 
> No, We currently have no plans on RNN as it's not very common nowadays. If you want peak performance, try https://github.com/intel/intel-extension-for-pytorch/tree/xpu-main

@jianyizh I have tried to roll back the pytorch version to 2.5.10+xpu and 2.3.110+xpu. In the same environment, running IMDB sentiment classification training (the code can be seen in the attachment in the initial bug description), it takes more than 40 seconds to complete an epoch under 2.6.10+xpu and 2.5.10+xpu, but only about 20 seconds under 2.3.110+xpu. Is this consistent with expectations?

**Comment by jianyizh** [comment]
> > > > RNN related OPs are not optimized on XPU backend, so it's expected to be slow
> > > 
> > > 
> > > I found an issue raised by you in the xpu-ops repo, which aims to integrate the oneDNN implementation of RNN and is in the pt2.8 milestone. Does that mean we may expect acceleration for RNNs in pytorch 2.8? Do we have other common NN achitectures that are not optimized in the xpu backend?
> > 
> > 
> > No, We currently have no plans on RNN as it's not very common nowadays. If you want peak performance, try https://github.com/intel/intel-extension-for-pytorch/tree/xpu-main
> 
> [@jianyizh](https://github.com/jianyizh) I have tried to roll back the pytorch version to 2.5.10+xpu and 2.3.110+xpu. In the same environment, running IMDB sentiment classification training (the code can be seen in the attachment in the initial bug description), it takes more than 40 seconds to complete an epoch under 2.6.10+xpu and 2.5.10+xpu, but only about 20 seconds under 2.3.110+xpu. Is this consistent with expectations?

Do you mean intel-extension-for-pytorch 2.6.10 is slower than intel-extension-for-pytorch 2.3? In this case, please raise an issue to them.

**Comment by chuanqi129** [comment]
Close it due to this is an Ipex issue
